src/persist-client/src/inspect.rs

```diff
@@ -324,6 +333,139 @@ pub async fn unreferenced_blobs(
+    loop {
+        let state_res = state_versions
+            .fetch_current_state::<K, V, u64, i64>(&shard_id, versions.clone())
+            .await;
+        match state_res {
+            Ok(_) => break,
+            Err(codec) => {
+                let mut kvtd = KVTD_CODECS.lock().expect("lockable");
+                *kvtd = codec.actual;
+            }
+        };
+    }
+
+    let mut machine = Machine::<K, V, u64, i64>::new(
+        cfg.clone(),
+        shard_id,
+        Arc::clone(&metrics),
+        Arc::clone(&state_versions),
+    )
+    .await?;
+
+    let writer_id = WriterId::new();
+    info!("registering writer {}", writer_id);
+    // We don't bother expiring this writer in various error codepaths, instead
+    // letting it time out. /shrug
+    let _ = machine
+        .register_writer(&writer_id, cfg.writer_lease_duration, (cfg.now)())
+        .await;
+
+    let mut attempt = 0;
+    'outer: loop {
+        machine.fetch_and_update_state().await;
+        let reqs = machine.state().collections.trace.all_fueled_merge_reqs();
+        info!("attempt {}: got {} compaction reqs", attempt, reqs.len());
+        for (idx, req) in reqs.clone().into_iter().enumerate() {
+            let req = CompactReq {
+                shard_id,
+                desc: req.desc,
+                inputs: req.inputs,
+            };
+            let parts = req.inputs.iter().map(|x| x.parts.len()).sum::<usize>();
+            let bytes = req
+                .inputs
+                .iter()
+                .flat_map(|x| x.parts.iter().map(|x| x.encoded_size_bytes))
+                .sum::<usize>();
+            let start = Instant::now();
+            info!(
+                "attempt {} req {}: compacting {} batches {} in parts {} totaling bytes: lower={:?} upper={:?} since={:?}",
+                attempt,
+                idx,
+                req.inputs.len(),
+                parts,
+                bytes,
+                req.desc.lower().elements(),
+                req.desc.upper().elements(),
+                req.desc.since().elements(),
+            );
+            let res = Compactor::<K, V, u64, i64>::compact(
+                cfg.clone(),
+                Arc::clone(&blob),
+                Arc::clone(&metrics),
+                Arc::new(CpuHeavyRuntime::new()),
+                req,
+                writer_id.clone(),
+            )
+            .await?;
+            info!(
+                "attempt {} req {}: compacted into {} parts {} bytes in {:?}",
+                attempt,
+                idx,
+                res.output.parts.len(),
+                res.output
+                    .parts
+                    .iter()
+                    .map(|x| x.encoded_size_bytes)
+                    .sum::<usize>(),
+                start.elapsed(),
+            );
+            let apply_res = machine
+                .merge_res(&FueledMergeRes { output: res.output })
```

Comment:
let's wrap this in the `if !read_only` check we plumb through

Reply:
👍 what you have here is what I had in mind, it's handy to have a no-op mode just to double check the work that would be done